src/engine/export/mmlOps.cpp

```diff
@@ -0,0 +1,320 @@
+      switch (cmd.cmd) {
+        case DIV_CMD_NOTE_ON: {
+          int noteVal = (chan == 3) ? _convertNoiseValue(cmd.value, useLegacyNoiseTable) : cmd.value;
+          st.currNote[chan] = noteVal;
+          st.noteOn[chan] = false;
+
+          auto ins = getIns(st.currIns[chan], DivInstrumentType::DIV_INS_GB);
+
+          if (chan != 2) {
+            if (!ins->gb.softEnv) {
+              // Initialize envelope if applicable
+              if ((ins->gb.alwaysInit || st.prevIns[chan] != st.currIns[chan]) && tick != st.volTick[chan]) {
+                st.currVol[chan] = ins->gb.envVol;
+              }
+              st.currEnv[chan] = ((ins->gb.envDir != 0) ? 1 : -1) * ins->gb.envLen;
+            }
+            // Set duty cycle for pulse channels or noise channel duty
+            if (chan < 2 && ins->std.dutyMacro.len > 0) {
+              st.currDuty[chan] = ins->std.dutyMacro.val[0];
+            } else if (chan == 3 && ins->std.dutyMacro.len > 0) {
+              st.currNoise = std::min(1, ins->std.dutyMacro.val[0]);
+            }
+          } else if (chan == 2 && ins->std.waveMacro.len > 0) {
+            st.currWave = ins->std.waveMacro.val[0];
+            if (st.currWave >= song.waveLen) st.currWave = song.waveLen - 1;
+          }
+
+          st.prevIns[chan] = st.currIns[chan];
+          break;
+        }
+        case DIV_CMD_NOTE_OFF:
+        case DIV_CMD_NOTE_OFF_ENV:
+          st.currNote[chan] = -1;
+          st.noteOn[chan] = false;
+          break;
+        case DIV_CMD_INSTRUMENT:
+          st.currIns[chan] = cmd.value;
+          break;
+        case DIV_CMD_HINT_VOLUME:
+          if (chan == 2) st.noteOn[chan] = false;
+          st.currVol[chan] = cmd.value;
+          st.volTick[chan] = tick;
+          break;
+        case DIV_CMD_PANNING: {
+          int l = cmd.value;
+          int r = cmd.value2;
+          if ((l == 0) == (r == 0)) st.currPan[chan] = 0;
+          else if (l == 0) st.currPan[chan] = 1;
+          else st.currPan[chan] = -1;
+          break;
+        }
+        case DIV_CMD_STD_NOISE_MODE:
+          if (chan < 2) {
+            st.currDuty[chan] = cmd.value;
+            st.dutyTick[chan] = tick;
+          } else if (chan == 3) {
+            st.currNoise = cmd.value;
+            st.noiseTick = tick;
+            st.noteOn[chan] = false;
+          }
+          break;
+        case DIV_CMD_WAVE:
+          if (chan == 2) {
+            st.currWave = cmd.value;
+            if (st.currWave >= song.waveLen) st.currWave = song.waveLen - 1;
+            st.waveTick = tick;
+            st.noteOn[chan] = false;
+          }
+          break;
+        default:
+          break;
+      }
+
+      // Handle pattern and measure boundaries for formatting new lines
+      if (curOrder != st.prevOrder[chan]) {
+        mmlStream += "\n\n" + cname + " ";
+        st.chanNameUsed[chan] = true;
+      } else if ((curRow - 1) / rowsPerMeasure != (st.prevRow[chan] - 1) / rowsPerMeasure) {
```

Comment:
Careful - possible division by zero. Don't perform this comparison if rowsPerMeasure is 0, or clamp rowsPerMeasure to 1 if less than 1.